In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


## Dataset Preprocessing

Working on dataset obtained from https://www.kaggle.com/c/jigsaw-toxic-comment-classification-challenge/data?select=train.csv.zip

This is a labelled dataset.

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv(r'/content/gdrive/MyDrive/train.csv')

In [ ]:
df['total'] = df['insult'] + df['threat'] + df['toxic'] + df['severe_toxic'] + df['obscene'] + df['identity_hate'] 

In [ ]:
df['total']

0         0
1         0
2         0
3         0
4         0
         ..
159566    0
159567    0
159568    0
159569    0
159570    0
Name: total, Length: 159571, dtype: int64

In [ ]:
# original dataset is highly biased
df['total'].value_counts()

0    143346
1      6360
3      4209
2      3480
4      1760
5       385
6        31
Name: total, dtype: int64

### Limiting the imbalance (all column labels = 0) to 25000 entries instead of 143346 for better results.

In [ ]:
count = 0
for i in range(len(df)):
  if df['total'][i] == 0 and count <118346:
    df = df.drop(axis = 0, index = i)
    count = count + 1
  if i % 1000 == 0:
    print(i)

0
1000
2000
3000
4000
5000
6000
7000
8000
9000
10000
11000
12000
13000
14000
15000
16000
17000
18000
19000
20000
21000
22000
23000
24000
25000
26000
27000
28000
29000
30000
31000
32000
33000
34000
35000
36000
37000
38000
39000
40000
41000
42000
43000
44000
45000
46000
47000
48000
49000
50000
51000
52000
53000
54000
55000
56000
57000
58000
59000
60000
61000
62000
63000
64000
65000
66000
67000
68000
69000
70000
71000
72000
73000
74000
75000
76000
77000
78000
79000
80000
81000
82000
83000
84000
85000
86000
87000
88000
89000
90000
91000
92000
93000
94000
95000
96000
97000
98000
99000
100000
101000
102000
103000
104000
105000
106000
107000
108000
109000
110000
111000
112000
113000
114000
115000
116000
117000
118000
119000
120000
121000
122000
123000
124000
125000
126000
127000
128000
129000
130000
131000
132000
133000
134000
135000
136000
137000
138000
139000
140000
141000
142000
143000
144000
145000
146000
147000
148000
149000
150000
151000
152000
153000
154000
155000
156000
157000
158000


In [ ]:
df

id  ... total
6       0002bcb3da6cb337  ...     4
12      0005c987bdfc9d4b  ...     1
16      0007e25b2121310b  ...     1
42      001810bf8c45bf5f  ...     4
43      00190820581d90ce  ...     3
...                  ...  ...   ...
159566  ffe987279560d7ff  ...     0
159567  ffea4adeee384e90  ...     0
159568  ffee36eab5c267c9  ...     0
159569  fff125370e4aaaf3  ...     0
159570  fff46fc426af1f9a  ...     0

[41225 rows x 9 columns]

In [ ]:
df['total'].value_counts()

0    25000
1     6360
3     4209
2     3480
4     1760
5      385
6       31
Name: total, dtype: int64

In [ ]:
df = df.reset_index(drop=True)

In [ ]:
df.to_csv('/content/gdrive/MyDrive/balanced_train.csv', index = False)

# Find embeddings

In [ ]:
!pip install sentence-transformers
from sentence_transformers import SentenceTransformer
roberta = SentenceTransformer('sentence-transformers/stsb-roberta-large')

     |████████████████████████████████| 79 kB 5.4 MB/s 
     |████████████████████████████████| 3.8 MB 43.2 MB/s 
     |████████████████████████████████| 1.2 MB 41.1 MB/s 
     |████████████████████████████████| 67 kB 4.9 MB/s 
     |████████████████████████████████| 596 kB 43.1 MB/s 
     |████████████████████████████████| 6.5 MB 33.2 MB/s 
     |████████████████████████████████| 895 kB 40.7 MB/s 
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.0-py3-none-any.whl size=120747 sha256=baaaaf9c5a4a918e8d7ee987162678d77d3cc5b526caef7036ec34e702d40bb3
  Stored in directory: /root/.cache/pip/wheels/83/c0/df/b6873ab7aac3f2465aa9144b6b4c41c4391cfecc027c8b07e7
Successfully built sentence-transformers
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


Downloading:   0%|          | 0.00/748 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/3.92k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.00 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/674 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/122 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/456k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/229 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/1.42G [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/52.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/239 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/1.17k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/798k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/191 [00:00<?, ?B/s]

In [ ]:
lst_embeddings = []

In [ ]:
import pandas as pd
df = pd.read_csv(r'/content/gdrive/MyDrive/balanced_train.csv')

In [ ]:
for idx in range(len(df['comment_text'])):
  if idx%1000==0:
    print(idx)
  sentences_embeddings = roberta.encode(df['comment_text'][idx])
  lst_embeddings.append(sentences_embeddings)

0
1000
2000
3000
4000
5000
6000
7000
8000
9000
10000
11000
12000
13000
14000
15000
16000
17000
18000
19000
20000
21000
22000
23000
24000
25000
26000
27000
28000
29000
30000
31000
32000
33000
34000
35000
36000
37000
38000
39000
40000
41000


In [ ]:
df['vectors'] = lst_embeddings

In [ ]:
df.to_csv('/content/gdrive/MyDrive/balanced_train_roberta_encodings.csv')

In [ ]:
df

id                                       comment_text  \
0      0002bcb3da6cb337       COCKSUCKER BEFORE YOU PISS AROUND ON MY WORK   
1      0005c987bdfc9d4b  Hey... what is it..\n@ | talk .\nWhat is it......   
2      0007e25b2121310b  Bye! \n\nDon't look, come or think of comming ...   
3      001810bf8c45bf5f  You are gay or antisemmitian? \n\nArchangel WH...   
4      00190820581d90ce           FUCK YOUR FILTHY MOTHER IN THE ASS, DRY!   
...                 ...                                                ...   
41220  ffe987279560d7ff  ":::::And for the second time of asking, when ...   
41221  ffea4adeee384e90  You should be ashamed of yourself \n\nThat is ...   
41222  ffee36eab5c267c9  Spitzer \n\nUmm, theres no actual article for ...   
41223  fff125370e4aaaf3  And it looks like it was actually you who put ...   
41224  fff46fc426af1f9a  "\nAnd ... I really don't think you understand...   

       toxic  severe_toxic  obscene  threat  insult  identity_hate  total  \
0          1             1        1       0       1              0      4   
1          1             0        0       0       0              0      1   
2          1             0        0       0       0              0      1   
3          1             0        1       0       1              1      4   
4          1             0        1       0       1              0      3   
...      ...           ...      ...     ...     ...            ...    ...   
41220      0             0        0       0       0              0      0   
41221      0             0        0       0       0              0      0   
41222      0             0        0       0       0              0      0   
41223      0             0        0       0       0              0      0   
41224      0             0        0       0       0              0      0   

                                                 vectors  
0      [0.13600183, -0.32565972, 0.50454044, 0.191750...  
1      [-0.91049415, -0.0044597, 0.8343384, 0.1426443...  
2      [-0.5363006, -0.33115152, -0.7755207, -0.03853...  
3      [0.49809888, -0.28926682, 0.6403113, -0.157657...  
4      [-0.559532, -0.10611109, -0.19170867, 0.070258...  
...                                                  ...  
41220  [0.6513231, -0.2367391, 0.47619265, 0.12619177...  
41221  [-0.3968309, 0.12859078, 1.299127, -0.4238446,...  
41222  [-0.07557843, -0.14591625, 0.18781255, 0.25630...  
41223  [-0.16929269, 0.17181423, 0.51042, -0.0807812,...  
41224  [0.78583163, 0.26408023, 0.2765813, 0.39752606...  

[41225 rows x 10 columns]

In [ ]:
import pandas as pd
import numpy as np
def converter(instr):
    return np.fromstring(instr[1:-1],sep=' ')
df1=pd.read_csv('/content/gdrive/MyDrive/balanced_train_roberta_encodings.csv',converters={'vectors':converter})


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: DeprecationWarning: string or file could not be read to its end due to unmatched data; this will raise a ValueError in the future.
  after removing the cwd from sys.path.


In [ ]:
type(df1['vectors'][1])

numpy.ndarray

## **Logistic Regression**

For target labels - 'toxic' and 'identity_hate'

In [ ]:
from sklearn.linear_model import LogisticRegression
model_identity_hate_lr = None
model_toxic_lr = None

import copy
def baseline_ml(target):
  X = df['vectors']
  #print(type(X[1]))
  y = df[target]
  print("Logistic Regression for - ",target)
  lst = []
  for i in X:
    lst.append(i)
  #print(type(lst[2]))

  from sklearn.model_selection import train_test_split
  x_train, x_test, y_train, y_test = train_test_split(lst, list(y), test_size=0.25, random_state=0)

  model = LogisticRegression(max_iter = 10000)
  model.fit(x_train, y_train)

  #Prediction using Logistic Reg
  if target == 'toxic':
    model_toxic_lr = copy.deepcopy(model)
    
  elif target == 'identity_hate':
    model_identity_hate_lr = copy.deepcopy(model)

  predictions = model.predict(x_test)
  
  #To improve recall
  # proba = model.predict_proba(x_test)
  # if target == 'toxic':
  #   model_toxic_lr = copy.deepcopy(model)
  # elif target == 'identity_hate':
  #   model_identity_hate_lr = copy.deepcopy(model)
  # predictions = []
  # for i in proba:
  #   if i[0] > 0.7:
  #     predictions.append(1)
  #   else:
  #     predictions.append(0)


  import matplotlib.pyplot as plt
  import seaborn as sns
  from sklearn import metrics
  cm = metrics.confusion_matrix(y_test, predictions)
  print("Confusion Matrix:")
  print(cm)
  precision = metrics.precision_score(y_test, predictions)
  recall = metrics.recall_score(y_test, predictions)
  F1 = 2 * (precision * recall) / (precision + recall)

  print("Accuracy of {} -- {}".format(target, metrics.accuracy_score(y_test, predictions)))
  print("Precision:",precision)
  print("Recall:",recall)
  print("F1 Score - ",F1)

  print("================")
  print()

  return model

In [ ]:
model_toxic_lr = baseline_ml('toxic')
model_identity_hate_lr = baseline_ml('identity_hate')

Logistic Regression for -  toxic
Confusion Matrix:
[[5963  485]
 [ 637 3222]]
Accuracy of toxic -- 0.8911419423692636
Precision: 0.8691664418667386
Recall: 0.8349313293599379
F1 Score -  0.851704996034893

Logistic Regression for -  identity_hate
Confusion Matrix:
[[9839  118]
 [ 220  130]]
Accuracy of identity_hate -- 0.967206752692345
Precision: 0.5241935483870968
Recall: 0.37142857142857144
F1 Score -  0.43478260869565216



## Predicting on unlabelled and unseen dataset

Tagged.csv - Data scraped from wikipedia(unbiased) and google(biased)

In [ ]:
df_tagged = pd.read_csv('/content/gdrive/MyDrive/tagged.csv')

In [ ]:
df_tagged
df_temp = pd.DataFrame()
l_s = []
l_pred = []
for i in range(len(df_tagged['Sentences'])):
  embs = roberta.encode(df_tagged['Sentences'][i])
  prediction = model_toxic_lr.predict(embs.reshape(1,-1))  
  l_s.append(df_tagged['Sentences'][i])
  l_pred.append(prediction[0])


In [ ]:
df_temp['sent'] = l_s
df_temp['preds'] = l_pred

In [ ]:
df_temp

sent  preds
0     'Real Housewives of Miami' reboot includes fra...      0
1     Reserve Bank Puts Off Plunging Country Into In...      0
2     A Latina scientist co-created a new Covid vacc...      0
3     Ayala: Democratic Latino voters may be agonizi...      0
4     India Defense Ministry denies clearance to fil...      0
...                                                 ...    ...
3613  The only way you're going to get Junior to vis...      0
3614  Sen. Kelly has apparently decided that the sen...      0
3615              What would you expect from this book?      0
3616  "And what this involved was a steady campaign ...      0
3617  And even more confusion today as he invokes th...      0

[3618 rows x 2 columns]

In [ ]:
df_temp.to_csv('/content/gdrive/MyDrive/predicted_roberta_toxicity.csv')

### Filtering out sentences that are predicted toxic in nature that is preds = 1.

In [ ]:
list_sentences = []
for i in range(len(df_temp['preds'])):
  if df_temp['preds'][i] == 1:
    list_sentences.append(df_temp['sent'][i])

In [ ]:
list_biased = ['gay','lesbian','black','transgender','chinese']

offensive_word_emb = roberta.encode(list_biased)

In [ ]:
lst_embeddings = []
for idx in range(len(df_temp['sent'])):
  if idx%1000==0:
    print(idx)
  sentences_embeddings = roberta.encode(df_temp['sent'][idx])
  lst_embeddings.append(sentences_embeddings)

0
1000
2000
3000


In [ ]:
len(lst_embeddings)

3618

In [ ]:
df_temp['vectors'] = lst_embeddings

In [ ]:
# performing custom tagging on test data sentences whose toxicity label was predicted as 1 by the logisitc regression model

list_sentences = []
list_vectors = []
for i in range(len(df_temp['preds'])):
  if df_temp['preds'][i] == 1:
    list_sentences.append(df_temp['sent'][i])
    list_vectors.append(df_temp['vectors'][i])

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

new_df = pd.DataFrame()
new_df['sentences'] = list_sentences
new_df['vectors'] = list_vectors
for i in list_biased:
  new_df[i] = ''
for each_vector in range(len(df_temp['vectors'])):
  for each_Word in range(len(offensive_word_emb)):
    cosine = cosine_similarity(offensive_word_emb[each_Word].reshape(1,-1), df_temp['vectors'][each_vector].reshape(1,-1))

    if cosine > 0.3:
      new_df[list_biased[each_Word]][each_vector] = 1
    else:
      new_df[list_biased[each_Word]][each_vector] = 0

In [ ]:
new_df

sentences  \
0                  Damn, he was a tank when he played.   
1    Pushing Party for Alex Morse Accountability, M...   
2    Worst Opinion of the Week: Lying Is Bad (Unles...   
3    Boston's Generation Z population want somewher...   
4    The Liberals instructed BC Hydro to hand over ...   
..                                                 ...   
543             What are These Boobies Doing in Maine?   
544  Doug Saunders,wannabe  member of the "PROGRESS...   
545  CIVIL COMMENTS BLOCKS FREE SPEECH\nCIVIL COMME...   
546  Man who threatened to kill Pelosi and Bowser p...   
547  MP’s wild 18 years: ‘Hunk of house’, tosser mo...   

                                               vectors gay lesbian black  \
0    [-0.6636137, 1.2023519, 0.6510567, 1.0746913, ...   0       0     0   
1    [-0.3408306, 0.9359817, 0.40372786, 1.9343902,...   0       0     0   
2    [-0.17794444, 1.4044044, 0.85383594, -0.849960...   0       0     0   
3    [-0.11416551, 0.91403675, -0.5913176, 1.057794...   0       0     0   
4    [-0.21968532, 1.7922213, 0.7958125, 0.5755419,...   0       0     0   
..                                                 ...  ..     ...   ...   
543  [-0.85967886, 0.81484836, 0.46285838, 1.052750...   0       0     0   
544  [0.4763507, 0.5380147, 0.83376664, -0.76617116...   0       0     0   
545  [-0.8941708, -0.28377637, 0.8854683, 0.7356403...   0       0     0   
546  [-0.5072312, 1.5499406, 0.46123874, -0.4379015...   0       0     0   
547  [-0.3076163, 0.28782687, -0.40475395, 1.087099...   0       0     0   

    transgender chinese  
0             0       0  
1             0       0  
2             0       0  
3             0       0  
4             0       0  
..          ...     ...  
543           0       0  
544           0       0  
545           0       0  
546           0       0  
547           0       0  

[548 rows x 7 columns]

In [ ]:
new_df.to_csv("/content/gdrive/MyDrive/roberta_custom_tagging_logistic.csv", index = False)